blockchain/chain_sync/src/network_context.rs

```diff
@@ -62,6 +65,47 @@ impl<DB: Clone> Clone for SyncNetworkContext<DB> {
     }
 }
 
+type RaceBatchFuture<T> = Pin<Box<dyn Future<Output = Result<T, String>> + Send>>;
+
+struct RaceBatch<T> {
+    tasks: JoinSet<Result<T, String>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl<T> RaceBatch<T>
+where
+    T: Send + 'static,
+{
+    pub fn new(max_concurrent_jobs: usize) -> Self {
+        RaceBatch {
+            tasks: JoinSet::new(),
+            semaphore: Arc::new(Semaphore::new(max_concurrent_jobs)),
+        }
+    }
+
+    pub fn add(&mut self, future: RaceBatchFuture<T>) {
+        let sem = self.semaphore.clone();
+        self.tasks.spawn(async move {
+            let permit = sem.acquire_owned().await.unwrap();
+            let result = future.await;
+            drop(permit);
```

Comment:
This slightly differs from the previous code as it doesn't guarantee that the futures are run in order. But it's OK because we've already shuffled the peers.